kuma/javascript/src/newsletter.jsx

```diff
@@ -89,68 +87,65 @@ export default function Newsletter() {
     const submit = event => {
         let newsletterForm = newsletterFormRef.current;
 
-        // if skipFetch
-        if (newsletterForm) {
-            if (newsletterForm.dataset['skipFetch']) {
-                /*
-                 * An error occured while attempting to subscribe
-                 * the user via Ajax, but no specific error was provided. We
-                 * therefore just send the user directly to the Mozorg
-                 * newsletter subscription page and log the occurence
-                 */
-                ga('send', {
-                    hitType: 'event',
-                    eventCategory: 'newsletter',
-                    eventAction: 'progression',
-                    eventLabel: 'error-forward'
-                });
-                return true;
-            }
-
-            event.preventDefault();
+        if (!newsletterForm) {
+            return;
+        }
 
-            let params = new URLSearchParams(
-                new FormData(newsletterForm)
-            ).toString();
+        // if skipFetch
+        if (newsletterForm.dataset['skipFetch']) {
+            /*
+             * An error occured while attempting to subscribe
+             * the user via Ajax, but no specific error was provided. We
+             * therefore just send the user directly to the Mozorg
+             * newsletter subscription page and log the occurence
+             */
+            ga('send', {
+                hitType: 'event',
+                eventCategory: 'newsletter',
+                eventAction: 'progression',
+                eventLabel: 'error-forward'
+            });
+            return true;
+        }
 
-            fetch(newsletterSubscribeURL, {
-                method: 'POST',
-                headers: {
-                    'X-Requested-With': 'XMLHttpRequest',
-                    'Content-type': 'application/x-www-form-urlencoded'
-                },
-                body: params
-            }).then(response => {
-                response.json().then(responseJSON => {
-                    if (responseJSON.success !== 'success') {
-                        let errorsArray = responseJSON.errors;
+        event.preventDefault();
+        let params = new URLSearchParams(
+            new FormData(newsletterForm)
+        ).toString();
+        fetch(newsletterSubscribeURL, {
+            method: 'POST',
+            headers: {
+                'X-Requested-With': 'XMLHttpRequest',
+                'Content-type': 'application/x-www-form-urlencoded'
+            },
+            body: params
+        })
+            .then(response => response.json())
```

Comment:
The response isn't always JSON. If `!response.ok` this'll fail cryptically.

Reply:
Good point, I've added an error handler

Reply:
That `.catch` will only happen if the request completely failed. For example, on network errors. The *communication* with the server might be successful but the response from the server might be `403 Forbidden` or `504 Bad Gateway`. That would not trigger a `.catch()` because `fetch` did as well as it possibly can. It's just a problem between the browser and the server. I often use this:

```javascript
let response
try {
   response = await fetch(url)
} catch(err) {
   return this.setState({networkError: err})
}
if (!response.ok) {
  return this.setState({serverError: response}) 
}
const data = await response.json();
return this.setState({ data });
```